platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java

```diff
@@ -905,27 +913,27 @@ public VirtualInternalRecord lookupInternalByPath(final long path, final boolean
      * <p>
      * This method may be called concurrently from multiple threads (although in practice, this should never happen).
      *
-     * @param firstLeafPath
-     * 		The first leaf path at and above which no internal node results should be returned. It is possible,
-     * 		through merging of multiple rounds, for the data to have internal data that is outside the expected range
+     * @param lastLeafPath
+     * 		The last leaf path at and above which no node results should be returned. It is possible,
+     * 		through merging of multiple rounds, for the data to have data that is outside the expected range
      * 		for the {@link VirtualMap} of this cache. We need to provide the leaf boundaries to compensate for this.
-     * @return A non-null stream of dirty internal records. May be empty. Will not contain duplicate records.
+     * @return A non-null stream of dirty records. May be empty. Will not contain duplicate records.
      * 		Will be sorted by path, with the lowest (root-most) path first.
      * @throws MutabilityException
      * 		if called on a non-sealed cache instance.
      */
-    public Stream<VirtualInternalRecord> dirtyInternals(long firstLeafPath) {
-        if (!dirtyInternals.isImmutable()) {
+    public Stream<PathHashRecord> dirtyHashes(final long lastLeafPath) {
+        if (!dirtyHashes.isImmutable()) {
             throw new MutabilityException("Cannot get the dirty internal records for a non-sealed cache.");
         }
 
-        final AtomicReference<Mutation<? extends VirtualRecord>> lastSeen = new AtomicReference<>();
-        return dirtyInternals
-                .sortedStream(dirtyInternalComparator())
-                .filter(mutation -> mutation.deleted || mutation.value.getPath() < firstLeafPath)
-                .filter(mutation -> dedupeByPath(mutation, lastSeen))
+        final AtomicReference<Mutation<Long, Hash>> lastSeen = new AtomicReference<>();
+        return dirtyHashes
+                .sortedStream(DIRTY_HASH_COMPARATOR)
+                .filter(mutation -> mutation.key <= lastLeafPath)
+                .filter(mutation -> dedupeHashByPath(mutation, lastSeen))
                 .filter(mutation -> !mutation.deleted)
-                .map(mutation -> mutation.value);
+                .map(mutation -> new PathHashRecord(mutation.key, mutation.value != NULL_HASH ? mutation.value : null));
```

Comment:
This will increase load on GC, need to check how it will impact performance.

Reply:
I checked the alternative and it looks like it would be worse. Alternatively, I could create `PathHashRecord` object in `VirtualNodeCache#updatePaths` method and keep it as a value in a `Mutation` object. However, it would mean that the path is stored twice - in `Mutation` as a key and in `PathHashRecord` that is stored as a value. 

In the current implementation `PathHashRecord` are short-lived as they don't escape `flush` method. So, there is a good chance that they will be collected by copying GC, and that will have no impact on the performance.